### PR TITLE
[WIP] Explicit Assignment Submission

### PR DESCRIPTION
**Pull request by cyhsutw**
Hello guys,

I'm working on the explicit submission feature and it's a one of the subtasks in #572.

The procedure of submission works as follows:

1. Classroom sets a organization webhook and receiving `push` events.
2. Upon receiving a `push` event of the **default branch**, Classroom will create a `release` for the `HEAD` commit.

Students can see their submissions in the release page of their repos:

releases

### TODO
- [ ] Delete all previous releases come with the starter code repo
- [ ] Ensure webhooks are correctly set: schedule background jobs to trigger `ping` events for hooks that have not been pinged
- [ ] Put this feature under Flipper
- [ ] More descriptive messages for the `release`s created by Classroom
- [ ] Add tests for this feature

### Questions / Discussions

- Since we're using `release` feature to create submissions, should we delete all the releases comes with the starter code repo? Yes.

- What do you think about exposing more information to the students so they can check if they have submitted their work properly?

    demo

### CODE REVIEW

**Comment by tarebyte** [comment]
> * Since we're using `release` feature to create submissions, should we delete all the releases comes with the starter code repo?

Yes we should, or we need to make a way to identify the releases we create.

**Comment by cyhsutw** [comment]
> Yes we should, or we need to make a way to identify the releases we create.

I'm 👍  with this.

Although currently we have a special prefix for submission releases: `classroom/submission/*` that can help us identify the submissions programmatically, deleting previous releases would make things more clear to the students (no submission at the point when the repo is created).

File: lib/classroom/scopes.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 module Classroom
   module Scopes
-    TEACHER                  = %w(user:email repo delete_repo admin:org).freeze
-    GROUP_ASSIGNMENT_STUDENT = %w(admin:org user:email).freeze
-    ASSIGNMENT_STUDENT       = %w(user:email).freeze
+    TEACHER                  = %w(user:email repo delete_repo admin:org admin:org_hook).freeze
```
Comment (cyhsutw) [inline lib/classroom/scopes.rb]: @tarebyte How can we make this Flippable? I think we don't have information about `current_user` here.

Comment (tarebyte) [inline lib/classroom/scopes.rb]: @cyhsutw I would just make a new constant, and then when we ask for the scope in the controller check for the `current_user` then.

File: app/controllers/organizations_controller.rb [hunk 2]
```diff
@@ -19,6 +19,9 @@ def new
 
   def create
     @organization = Organization.new(new_organization_params)
+    if Classroom.flipper[:explicit_assignment_submission].enabled? current_user
+      @organization.setup_webhook(webhook_events_url)
```
Comment (cyhsutw) [inline app/controllers/organizations_controller.rb]: @tarebyte I was doing this in the `before_validation(on: :create)` callback of the `Organization` model. However, as we must have the `webhook_events_url` in order to setup the webhook, including `Rails.application.routes.url_helpers` in `Organization` and setting the `host` of the application in `environment/*.rb` are necessary.

Is it fine to place this piece of code here or we have better options?

Comment (tarebyte) [inline app/controllers/organizations_controller.rb]: @cyhsutw We should probably make it a service object, I'd really rather not put the logic for this in a controller.

Comment (cyhsutw) [inline app/controllers/organizations_controller.rb]: Thanks for the suggestion, I'll try to do that.

**Comment by cyhsutw** [comment]
Close for a better PR.

**Pull request closed without merging**